### PR TITLE
Performance improvements and other updates

### PR DESCRIPTION
**Pull request by devmotion**
Hej!

Thanks for this package! I made some changes that seem to speed up the computations on my computer at least for a simple example. Moreover, I switched from `REQUIRE` to `Project.toml`. Adding the same compatibilities in the `[compat]` format to `Project.toml` broke dependencies for me, so I only added a bound for the Julia version.

I ran the following script:
```julia
using BenchmarkTools
using Bootstrap
using Bootstrap.Datasets
using DataFrames
using Statistics

city_ratio(x::AbstractDataFrame) = mean(x[:X]) ./ mean(x[:U])
city_cor(x::AbstractDataFrame) = cor(x[:X], x[:U])

function bootstrap_ci(statistic, data, sampling, cim)
    bs = bootstrap(statistic, data, sampling)
    c = confint(bs, cim)
end

function run_benchmark()
    bs = BasicSampling(1_000)
    cim = BasicConfInt()

    @btime bootstrap_ci($(city_ratio), $(city), $(bs), $(cim))
    @btime bootstrap_ci($(city_cor), $(city), $(bs), $(cim))
end
```

On the master branch I got:
```julia
julia> run_benchmark()
  1.625 ms (21553 allocations: 1.03 MiB)
  1.276 ms (17549 allocations: 964.34 KiB)
```
And with this PR:
```julia
julia> run_benchmark()
  1.194 ms (18534 allocations: 916.83 KiB)
  891.029 μs (14528 allocations: 822.95 KiB)
```

### CODE REVIEW

**Comment by juliangehring** [comment]
Thanks, that looks very interesting! I only had a quick look over the changes and I might not find time to review them properly for a few days.

Can you rebase your pull-request against the `develop` branch? We only use the master branch for stable releases (I need to document this properly... see #44). The changes to `Project.toml` and friends already landed there with #48.

File: src/bootsampling.jl [hunk 1]
```diff
@@ -177,17 +177,21 @@ noise(bs::WildSampling) = bs.noise
...
+_zeros_tuple(::Type{Tuple{}}, m::Int) = ()
+_zeros_tuple(::Type{T}, m::Int) where T <: Tuple =
+    (zeros(Base.tuple_type_head(T), m), _zeros_tuple(Base.tuple_type_tail(T), m)...)
 
 lhs(f::Formula) = f.lhs
 
 """
 bootstrap(statistic, data, BasicSampling())
 """
-function bootstrap(statistic::Function, data, sampling::BasicSampling)
```
Comment (juliangehring) [inline src/bootsampling.jl]: Why do we need to drop `Function` here? Is it too restrictive?

File: src/confint.jl [hunk 2]
```diff
@@ -10,8 +10,8 @@ BasicConfInt(0.95)
 ```
 
 """
-struct BasicConfInt <: ConfIntMethod
```
Comment (juliangehring) [inline src/confint.jl]: Does switching to a parametrised method improve the performance or is this change rather aesthetically?

File: src/draw.jl [hunk 3]
```diff
@@ -1,6 +1,6 @@
 ## draw: unify rand, sample
 
-draw!(x::T, o) where {T<:Distribution} = rand!(x, o)
```
Comment (juliangehring) [inline src/draw.jl]: Is this related to performance or aesthetics?

**Comment by devmotion** [comment]
Ah, somehow I missed the `develop` branch and the existing changes there. I'll try to rebase the changes against this branch.

File: src/bootsampling.jl (hunk 1 above)
Comment (devmotion) [reply src/bootsampling.jl]: Yes, in my opinion it's too restrictive. It rules out the common use case of [function-like objects](https://docs.julialang.org/en/v1/manual/methods/#Function-like-objects-1).

File: src/confint.jl (hunk 2 above)
Comment (devmotion) [reply src/confint.jl]: I haven't benchmarked every change in detail but in general type declarations should [improve the performance](https://docs.julialang.org/en/v1/manual/performance-tips/index.html#Type-declarations-1).

File: src/draw.jl (hunk 3 above)
Comment (devmotion) [reply src/draw.jl]: There's no performance difference but it's [recommended to remove unnecessary static parameters](https://docs.julialang.org/en/v1/manual/style-guide/#Don't-use-unnecessary-static-parameters-1).

File: src/confint.jl (hunk 2 above)
Comment (rofinn) [reply src/confint.jl]: NOTE: Parameterizing the type only improves performance if you're dealing with wrapper types (e.g., `AbstractFloat`). If `level` was declared as `Float64` then this wouldn't make a difference.

File: src/bootsampling.jl [hunk 4]
```diff
@@ -114,30 +114,30 @@ maximumEntropySampling(100, MaximumEntropyCache())
 NOTE: Implementation based off [pymeboot](https://github.com/kirajcg/pymeboot) as the original
 [R package](https://cran.r-project.org/web/packages/meboot/index.html) is GPL licensed.
 """
-struct MaximumEntropySampling <: BootstrapSampling
+struct MaximumEntropySampling{C<:MaximumEntropyCache} <: BootstrapSampling
     nrun::Int
-    cache::MaximumEntropyCache
+    cache::C
```
Comment (rofinn) [inline src/bootsampling.jl]: This doesn't need to parameterized because `MaximumEntropyCache` is a concrete type (e.g., not like `AbstractFloat`).

File: src/draw.jl [hunk 5]
```diff
@@ -72,15 +70,14 @@ function MaximumEntropyCache()
     )
 end
 
-function init!(cache::MaximumEntropyCache, x::AbstractArray)
+function init!(cache::MaximumEntropyCache{T}, x::AbstractArray) where T
```
Comment (rofinn) [inline src/draw.jl]: 👍 This seems like a good change. I should have noticed this earlier :)

File: src/bootsampling.jl (hunk 4 above)
Comment (devmotion) [reply src/bootsampling.jl]: No, since the type parameter is not declared it's ambiguous. Like the example of `Point` in the [documentation](https://docs.julialang.org/en/v1/manual/types/index.html#Parametric-Composite-Types-1), the type object `MaximumEntropyCache` contains all instances `MaximumEntropyCache{Float64}`,`MaximumEntropyCache{Integer}`,etc. as subtypes.

File: src/confint.jl (hunk 2 above)
Comment (devmotion) [reply src/confint.jl]: Are you referring to my comment? Defining `level` as a field with a concrete type like `Float64` would improve the performance as well but, of course, be much more restrictive. Currently (on the master and develop branch) the implementation of `BasicConfInt` is identical to the example `MyStillAmbiguousType` in the [discussion about fields with abstract type](https://docs.julialang.org/en/v1/manual/performance-tips/index.html#Avoid-fields-with-abstract-type-1).

Comment (devmotion) [reply src/confint.jl]: But as the example in the documentation shows, just constructing `BasicConfInt` with a concrete type such as `BasicConfInt(0.95)` is problematic.

**Comment by devmotion** [comment]
Just a quick thing I tried:
```julia
using Bootstrap
using BenchmarkTools
using StatsBase

function test_mean_and_std(n)
    r = rand(25)
    sampling = BasicSampling(n)

    @btime bootstrap($mean_and_std, $r, $sampling)
end

function test_bca_conf_int(n)
    sample = bootstrap(mean_and_std, rand(25), BasicSampling(n))
    ci = BCaConfInt(0.95)

    @btime confint($sample, $ci)
end
```

On the current master I get:
```julia
julia> test_mean_and_std(1000);
  296.483 μs (8 allocations: 16.38 KiB)
julia> test_bca_conf_int(1000);
  163.696 μs (349 allocations: 47.53 KiB)
```
On the branch dev (this PR):
```julia
julia> test_mean_and_std(1000);
  291.477 μs (5 allocations: 16.23 KiB)
julia> test_bca_conf_int(1000);
  131.437 μs (238 allocations: 44.97 KiB)
```
In more detail, with https://github.com/juliangehring/Bootstrap.jl/pull/49/commits/99649539c8ff673b03aed6e88f0d2d9cc4661544 ("Performance enhancements") I get:
```julia
julia> test_mean_and_std(1000);
  293.313 μs (7 allocations: 16.28 KiB)
julia> test_bca_conf_int(1000);
  145.351 μs (344 allocations: 47.28 KiB)
```
With https://github.com/juliangehring/Bootstrap.jl/pull/49/commits/d90e66cde6f4111c560467bbe90e0642d8a62296 ("Remove Function annotation"):
```julia
julia> test_mean_and_std(1000);
  292.425 μs (7 allocations: 16.28 KiB)
julia> test_bca_conf_int(1000);
  146.695 μs (344 allocations: 47.28 KiB)
```
With https://github.com/juliangehring/Bootstrap.jl/pull/49/commits/fab92a19ddf1a71b9653f3d52f7c67827f46d907 ("Add type parameters"):
```julia
julia> test_mean_and_std(1000);
  292.907 μs (5 allocations: 16.23 KiB)
julia> test_bca_conf_int(1000);
  118.622 μs (238 allocations: 44.97 KiB)
```

To me that clearly shows that the additional type parameters are helpful :smiley:

**Comment by rofinn** [comment]
> To me that clearly shows that the additional type parameters are helpful

True, but I wonder how much of that is just from not storing an `AbstractFloat` for the `ConfIntMethod`s. It looks like most of the improvement is only on calling `confint`, so parameterizing the sampling methods and samples may be excessive?

File: src/confint.jl (hunk 2 above)
Comment (rofinn) [reply src/confint.jl]: > just constructing BasicConfInt with a concrete type such as BasicConfInt(0.95) is problematic.

I cleared missed something, is there some documentation that suggests we might want to use Float32?

File: src/bootsampling.jl (hunk 4 above)
Comment (rofinn) [reply src/bootsampling.jl]: Oh, yeah. FWIW, I'm not sure how much this'll get us because the cache contains a bunch of container objects, but I guess it doesn't hurt. I'd probably do this though.

```
struct MaximumEntropySampling{T<:Real}
    nrun::Int
    cache::MaximumEntropyCache{T}
end
```

File: src/confint.jl (hunk 2 above)
Comment (devmotion) [reply src/confint.jl]: > I clearly missed something, is there some documentation that suggests we might want to use Float32?

Sorry, I don't understand your question. What do you mean? Would you like to define
```julia
struct BasicConfInt
     level::Float64
end
```
instead and are wondering why that would be restrictive? My point was that this would be much more restrictive than the current declaration `level::AbstractFloat`. As you say, someone might want to use `Float32` (e.g., it's the standard floating point number type on Windows 32bit, if I'm not mistaken), dual numbers, or some arbitrary user-defined number types. I'm not sure whether people would make use of that to a large extent, but when making the changes in this PR, defining
```julia
struct BasicConfInt{T<:AbstractFloat}
     level::T
end
```
was the only (?) way to get a composite type without abstract fields that still gives the user the same amount of freedom and flexibility in providing different types of `level` as in the current implementation on the master branch.

**Comment by devmotion** [comment]
> It looks like most of the improvement is only on calling confint, so parameterizing the sampling methods and samples may be excessive?

In my opinion it's not. Basically, as stated in the documentation, due to the unspecified types the type inference of the compiler breaks down which leads to suboptimal machine code and hence subpar performance not only when calling `confint` but in any operations that use these sample types. Look at the following simple example:
```julia
function original_op(n)
    sample = bootstrap(mean_and_std, rand(25), BasicSampling(n))
    10 * first(original(sample))
end
```
Without type parameters (on https://github.com/juliangehring/Bootstrap.jl/commit/d90e66cde6f4111c560467bbe90e0642d8a62296) the compiler is not able to infer the return type of this simple function and not able to optimize the last multiplication operation:
```julia
julia> @code_warntype original_op(1000);
Body::Any
1 ── %1  = Random.GLOBAL_RNG::Random.MersenneTwister
│    %2  = $(Expr(:foreigncall, :(:jl_alloc_array_1d), Array{Float64,1}, svec(Any, Int64), :(:ccall), 2, Array{Float64,1}, 25, 25))::Array{Float64,1}
│    %3  = (Base.arraylen)(%2)::Int64
│    %4  = (Base.mul_int)(8, %3)::Int64
│    %5  = (Base.arraylen)(%2)::Int64
│    %6  = (Base.mul_int)(8, %5)::Int64
│    %7  = (Base.sle_int)(%4, %6)::Bool
└───       goto #3 if not %7
2 ──       goto #4
3 ──       nothing
4 ┄─ %11 = φ (#2 => true, #3 => false)::Bool
└───       goto #6 if not %11
5 ── %13 = $(Expr(:gc_preserve_begin, :(%2)))
│    %14 = $(Expr(:foreigncall, :(:jl_array_ptr), Ptr{Float64}, svec(Any), :(:ccall), 1, :(%2)))::Ptr{Float64}
│    %15 = %new(Random.UnsafeView{Float64}, %14, %3)::Random.UnsafeView{Float64}
│          invoke Random.rand!(%1::Random.MersenneTwister, %15::Random.UnsafeView{Float64}, $(QuoteNode(Random.SamplerTrivial{Random.CloseOpen01{Float64},Float64}(Random.CloseOpen01{Float64}())))::Random.SamplerTrivial{Random.CloseOpen01{Float64},Float64})
│          $(Expr(:gc_preserve_end, :(%13)))
└───       goto #7
6 ── %19 = %new(Core.AssertionError, "sizeof(Float64) * n64 <= sizeof(T) * length(A) && isbitstype(T)")::AssertionError
│          (Base.throw)(%19)
└───       $(Expr(:unreachable))
7 ┄─       goto #8
8 ──       goto #9
9 ──       goto #10
10 ─       goto #11
11 ─       goto #12
12 ─ %27 = %new(Bootstrap.BasicSampling, n)::BasicSampling
│    %28 = invoke Main.bootstrap(Main.mean_and_std::typeof(mean_and_std), %2::Array{Float64,1}, %27::BasicSampling)::NonParametricBootstrapSample{Array{Float64,1}}
│    %29 = (Base.getfield)(%28, :t0)::Tuple
│    %30 = Main.first::Core.Compiler.Const(first, false)
│    %31 = (isa)(%29, Tuple{})::Bool
└───       goto #14 if not %31
13 ─ %33 = π (%29, Tuple{})
│    %34 = invoke %30(%33::Tuple{})::Any
└───       goto #15
14 ─ %36 = (Main.first)(%29)::Any
└───       goto #15
15 ┄ %38 = φ (#13 => %34, #14 => %36)::Any
│    %39 = (10 * %38)::Any
└───       return %39
```
However, with type parameters (on https://github.com/juliangehring/Bootstrap.jl/commit/fab92a19ddf1a71b9653f3d52f7c67827f46d907) the return type is inferred and the final operations are optimized for the return type of the sample statistic (compare the last lines of the outputs):
```julia
julia> @code_warntype original_op(1000);
Body::Float64
1 ── %1  = Random.GLOBAL_RNG::Random.MersenneTwister
│    %2  = $(Expr(:foreigncall, :(:jl_alloc_array_1d), Array{Float64,1}, svec(Any, Int64), :(:ccall), 2, Array{Float64,1}, 25, 25))::Array{Float64,1}
│    %3  = (Base.arraylen)(%2)::Int64
│    %4  = (Base.mul_int)(8, %3)::Int64
│    %5  = (Base.arraylen)(%2)::Int64
│    %6  = (Base.mul_int)(8, %5)::Int64
│    %7  = (Base.sle_int)(%4, %6)::Bool
└───       goto #3 if not %7
2 ──       goto #4
3 ──       nothing
4 ┄─ %11 = φ (#2 => true, #3 => false)::Bool
└───       goto #6 if not %11
5 ── %13 = $(Expr(:gc_preserve_begin, :(%2)))
│    %14 = $(Expr(:foreigncall, :(:jl_array_ptr), Ptr{Float64}, svec(Any), :(:ccall), 1, :(%2)))::Ptr{Float64}
│    %15 = %new(Random.UnsafeView{Float64}, %14, %3)::Random.UnsafeView{Float64}
│          invoke Random.rand!(%1::Random.MersenneTwister, %15::Random.UnsafeView{Float64}, $(QuoteNode(Random.SamplerTrivial{Random.CloseOpen01{Float64},Float64}(Random.CloseOpen01{Float64}())))::Random.SamplerTrivial{Random.CloseOpen01{Float64},Float64})
│          $(Expr(:gc_preserve_end, :(%13)))
└───       goto #7
6 ── %19 = %new(Core.AssertionError, "sizeof(Float64) * n64 <= sizeof(T) * length(A) && isbitstype(T)")::AssertionError
│          (Base.throw)(%19)
└───       $(Expr(:unreachable))
7 ┄─       goto #8
8 ──       goto #9
9 ──       goto #10
10 ─       goto #11
11 ─       goto #12
12 ─ %27 = %new(Bootstrap.BasicSampling, n)::BasicSampling
│    %28 = invoke Main.bootstrap(Main.mean_and_std::typeof(mean_and_std), %2::Array{Float64,1}, %27::BasicSampling)::NonParametricBootstrapSample{Array{Float64,1},Tuple{Float64,Float64},Tuple{Array{Float64,1},Array{Float64,1}},typeof(mean_and_std),BasicSampling}
│    %29 = (Base.getfield)(%28, :t0)::Tuple{Float64,Float64}
│    %30 = (Base.getfield)(%29, 1, true)::Float64
│    %31 = (Base.mul_float)(10.0, %30)::Float64
└───       return %31
``` 
As expected, this difference is reflected in the higher number of allocations and slower run time without type parameters
```julia
julia> test_original_op(1000);
  273.427 μs (10 allocations: 16.59 KiB)
```
compared with the implementation with type parameters:
```julia
julia> test_original_op(1000);
  265.814 μs (6 allocations: 16.52 KiB)
```

**Comment by devmotion** [comment]
What's your current opinion about the changes in this PR? Would you like me to open a clean new PR (with a subset of those maybe?) that accounts for the changes in the develop branch?

**Comment by juliangehring** [comment]
@devmotion I'd be happy to have a closer look at a clean PR against the current `develop` branch with the performance improvements.

**Comment by juliangehring** [comment]
Closing since the relevant changes have now been merged with #66 and #67. Thanks to everyone for their ideas and comments.
